torch/csrc/api/include/torch/nn/modules/rnn.h

```diff
@@ -40,62 +42,83 @@ class RNNImplBase : public torch::nn::Cloneable<Derived> {
+  /// parameter is modified__, e.g. any time a parameter is assigned a new
+  /// value. This allows using the fast path in cuDNN implementations of
+  /// respective RNN `forward()` methods. It is called once upon construction,
+  /// inside `reset()`.
+  void flatten_parameters();
+
+  /// Returns the cuDNN mode of the RNN subclass if it has one. See
+  /// https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnRNNMode_t
+  /// for a list of all cuDNN modes.
+  at::optional<CuDNNMode> cudnn_mode() const noexcept;
 
   RNNOptionsBase options;
 
+  /// The weights for `input x hidden` gates.
   std::vector<Tensor> w_ih;
+  /// The weights for `hidden x hidden` gates.
   std::vector<Tensor> w_hh;
+  /// The biases for `input x hidden` gates.
   std::vector<Tensor> b_ih;
+  /// The biases for `hidden x hidden` gates.
   std::vector<Tensor> b_hh;
 
-  Dropout dropout;
+  /// The dropout module, if dropout is used.
+  Dropout dropout{nullptr};
 
  protected:
-  virtual Tensor cell_forward(Tensor input, Tensor state, int64_t layer) = 0;
-
-  RNNOutput CUDNN_forward(Tensor input, Tensor state);
-  RNNOutput autograd_forward(Tensor input, Tensor state);
-
+  /// The function signature of `at::lstm`, `at::rnn_relu`, `at::gru` etc.
+  using RNNFunctionSignature = std::tuple<Tensor, Tensor>(
+      /*input=*/const Tensor&,
+      /*state=*/const Tensor&,
+      /*params=*/TensorList,
+      /*has_biases=*/bool,
+      /*layers=*/int64_t,
+      /*dropout=*/double,
+      /*train=*/bool,
+      /*bidirectional=*/bool,
+      /*batch_first=*/bool);
+
+  RNNOutput generic_forward(
+      std::function<RNNFunctionSignature> function,
+      Tensor input,
+      Tensor state);
+
+  /// Returns a flat vector of all weights, with layer weights following each
+  /// other sequentially in (w_ih, w_hh, b_ih, b_hh) order.
   std::vector<Tensor> flat_weights() const;
-  bool use_cudnn(Tensor sample) const;
-  Tensor create_dropout_state(Tensor input) const;
 
+  /// Returns true if any of the parameters (weights, biases) alias each other.
+  bool any_parameters_alias() const;
+
+  /// The number of gate weights/biases required by the RNN subclass.
   int64_t number_of_gates_;
-  bool has_cell_state_;
+
+  /// The cuDNN RNN mode, if this RNN subclass has any.
   at::optional<CuDNNMode> cudnn_mode_;
 
-  // This is copied from pytorch, to determine whether weights are flat for the
-  // fast CUDNN route. Otherwise, we have to use non flattened weights, which
-  // are much slower.
-  // https://github.com/pytorch/pytorch/blob/1848cad10802db9fa0aa066d9de195958120d863/torch/nn/modules/rnn.py#L159-L165
-  // TODO Actually since we are in C++ we can probably just actually check if
-  // the parameters are flat, instead of relying on data pointers and stuff.
-  std::vector<void*> data_ptrs_;
-  Tensor flat_weights_;
+  /// The cached result of the latest `flat_weights()` call.
+  std::vector<Tensor> flat_weights_;
```

Comment:
But it will be super slow if it's not flattened no?

Reply:
Also `flat_weights_` here is more about having the weights layed out in the way `at::lstm` et al. expect them, i.e. `(w_ih, w_hh, b_ih, b_hh)`

Reply:
Oh I read that as an at::Tensor that corresponds to their flattening according to cuDNN layout. I guess this is fine, but I feel a bit bad about caching those, because invalidation might be hard.